private/bufpkg/bufwasm/bufwasm.go

```diff
@@ -149,27 +149,36 @@ func WithMemoryLimitPages(memoryLimitPages uint32) PluginExecutorOption {
 // CompilePlugin takes a byte slice with a valid wasm module, compiles it and
 // optionally reads out buf plugin metadata.
 func (e *WASMPluginExecutor) CompilePlugin(ctx context.Context, plugin []byte) (_ *CompiledPlugin, retErr error) {
-
 	// Configure the compilation cache, which must be closed after the runtime.
 	var cache wazero.CompilationCache
 	if e.compilationCacheDir == "" {
 		cache = wazero.NewCompilationCache()
 	} else {
-		cache, retErr = wazero.NewCompilationCacheWithDir(e.compilationCacheDir)
-		if retErr != nil {
-			return
+		var err error
+		cache, err = wazero.NewCompilationCacheWithDir(e.compilationCacheDir)
+		if err != nil {
+			return nil, err
 		}
 	}
+	defer func() {
+		if retErr != nil {
+			retErr = multierr.Append(retErr, cache.Close(ctx))
+		}
+	}()
 
 	// Create the shared runtime used for all plugin instantiations
 	runtime := wazero.NewRuntimeWithConfig(ctx, e.runtimeConfig.WithCompilationCache(cache))
+	defer func() {
+		if retErr != nil {
+			retErr = multierr.Append(retErr, runtime.Close(ctx))
+		}
+	}()
 
 	// Note: before we start accepting user plugins, we should do more
 	// validation on the metadata here: file path cleaning etc.
-	var compiledModule wazero.CompiledModule
-	compiledModule, retErr = runtime.CompileModule(ctx, plugin)
```

Comment:
Small style tweak here - we typically only use the named return variables in a defer.